File: apps/web/types/technical.ts
export interface EMASnapshot {
  ema_20: number | null;
  ema_50: number | null;
  ema_200: number | null;
}

export interface SMASnapshot {
  sma_20: number | null;
  sma_50: number | null;
  sma_200: number | null;
}

export interface MACDSnapshot {
  macd: number | null;
  signal: number | null;
  histogram: number | null;
}

export interface BollingerBandsSnapshot {
  high_band: number | null;
  mid_band: number | null;
  low_band: number | null;
}

export interface PivotPointsSnapshot {
  pp: number | null;
  r1: number | null;
  r2: number | null;
  r3: number | null;
  s1: number | null;
  s2: number | null;
  s3: number | null;
}

export interface IndicatorSnapshot {
  ticker: string;
  updated_at: string;
  ema: EMASnapshot;
  sma: SMASnapshot;
  macd: MACDSnapshot;
  adx: number | null;
  rsi: number | null;
  mfi: number | null;
  atr: number | null;
  bollinger_bands: BollingerBandsSnapshot;
  vwap: number | null;
  obv: number | null;
  cmf: number | null;
  pivot_points: PivotPointsSnapshot;
}

export type SignalType = "STRONG_BUY" | "BUY" | "NEUTRAL" | "SELL" | "STRONG_SELL";
export type IndicatorSignalType = "BUY" | "SELL" | "NEUTRAL";
export type SignalCategory = "Trend" | "Momentum" | "Volatility" | "Volume" | "Support/Resistance";
export type SignalStrength = "High" | "Medium" | "Low";

export interface TechnicalSignal {
  name: string;
  category: SignalCategory;
  indicator: string;
  signal: IndicatorSignalType;
  strength: SignalStrength;
  value: number | null;
  description: string;
  timestamp: string;
}

export interface TechnicalSummary {
  trend: "Bullish" | "Bearish" | "Neutral";
  momentum: "Bullish" | "Bearish" | "Neutral";
  volume: "High" | "Average" | "Low";
  volatility: "High" | "Average" | "Low";
  support: number | null;
  resistance: number | null;
  overall_rating: SignalType;
}

export interface SignalsSnapshot {
  ticker: string;
  updated_at: string;
  summary: TechnicalSummary;
  signals: TechnicalSignal[];
}

File: apps/api/app/schemas/technical.py
from typing import Literal
from pydantic import BaseModel, Field

class EMASnapshot(BaseModel):
    ema_20: float | None = None
    ema_50: float | None = None
    ema_200: float | None = None

class SMASnapshot(BaseModel):
    sma_20: float | None = None
    sma_50: float | None = None
    sma_200: float | None = None

class MACDSnapshot(BaseModel):
    macd: float | None = None
    signal: float | None = None
    histogram: float | None = None

class BollingerBandsSnapshot(BaseModel):
    high_band: float | None = None
    mid_band: float | None = None
    low_band: float | None = None

class PivotPointsSnapshot(BaseModel):
    pp: float | None = None
    r1: float | None = None
    r2: float | None = None
    r3: float | None = None
    s1: float | None = None
    s2: float | None = None
    s3: float | None = None

class IndicatorSnapshot(BaseModel):
    ticker: str
    updated_at: str
    
    # Trend Indicators
    ema: EMASnapshot
    sma: SMASnapshot
    macd: MACDSnapshot
    adx: float | None = None
    
    # Momentum Indicators
    rsi: float | None = None
    mfi: float | None = None
    
    # Volatility Indicators
    atr: float | None = None
    bollinger_bands: BollingerBandsSnapshot
    
    # Volume Indicators
    vwap: float | None = None
    obv: float | None = None
    cmf: float | None = None
    
    # Support/Resistance
    pivot_points: PivotPointsSnapshot

class TechnicalSignal(BaseModel):
    name: str
    category: Literal["Trend", "Momentum", "Volatility", "Volume", "Support/Resistance"]
    indicator: str
    signal: Literal["BUY", "SELL", "NEUTRAL"]
    strength: Literal["High", "Medium", "Low"]
    value: float | None = None
    description: str
    timestamp: str

class TechnicalSummary(BaseModel):
    trend: Literal["Bullish", "Bearish", "Neutral"]
    momentum: Literal["Bullish", "Bearish", "Neutral"]
    volume: Literal["High", "Average", "Low"]
    volatility: Literal["High", "Average", "Low"]
    support: float | None = None
    resistance: float | None = None
    overall_rating: Literal["STRONG_BUY", "BUY", "NEUTRAL", "SELL", "STRONG_SELL"]

class SignalsSnapshot(BaseModel):
    ticker: str
    updated_at: str
    summary: TechnicalSummary
    signals: list[TechnicalSignal]
